src/printer.js

```diff
@@ -1746,7 +1746,11 @@ function genericPrintNoParens(path, options, print) {
     case "NumberLiteralTypeAnnotation":
       assert.strictEqual(typeof n.value, "number");
 
-      return "" + n.value;
+      if (n.extra != null) {
+        return printNumber(n.extra.raw);
+      } else {
+        return printNumber(n.raw);
+      }
```

Comment:
Running the existing tests shows that sometimes `n.extra` is undefined, sometimes not, hence the two cases here.